Remove "old" signature spoofing patches for LineageOS 21, and introduce patch which enables it even in the user build variant

nnsee opened the pull request:
As discussed in https://github.com/lineageos4microg/docker-lineage-cicd/issues/607, this PR removes the signature spoofing patches entirely for lineage-21.0 (Android 14), since 21 has signature spoofing functionality built in for `eng` and `userdebug` builds (even with `SIGNATURE_SPOOFING` set to `no`).

Instead, this PR replaces the standard patch with one which enables signature spoofing even for `user` builds if `SIGNATURE_SPOOFING` is set to either `restricted` or `yes` (both do the same thing).

Since the built-in signature spoofing checks whether the app is microG specifically, it is no longer possible to allow signature spoofing to other third party apps. If it turns out that this is a use case which is required, the "old" signature spoofing patches need to be fixed in order to build on LineageOS 21 and reintroduced (alongside this patch, so users have the option to use both if desired).

FintasticMan commented on Dockerfile:
```diff
@@ -94,6 +94,14 @@ ENV ZIP_SUBDIR true
+# with signature spoofing enabled; otherwise, it will be disabled in user
+# builds.
+#
+# For LineageOS 20 and older:
```
LOS has added signature spoofing to 18.1, 19.1, 20.0 and 21.0. L4M's patches just aren't broken on versions below 21 AFAIK.

nnsee replied:
Oh, I didn't realize older LOS versions also now have built-in signature spoofing. Will change documentation to reflect this.

FintasticMan replied:
You could also add the same patches for removing the debug check for those versions, the code should be the same.

FintasticMan reviewed (commented):
I think it would make most sense to add a separate env var for disabling the debug build check, as it is very distinct from the "old" patches.

nnsee commented:
> I think it would make most sense to add a separate env var for disabling the debug build check, as it is very distinct from the "old" patches.

This makes sense, will change.

petefoth commented:
The current logic, for all build versions from 18.1 onwards, is that:
1. The LOS changes are applied, in `userdebug` builds only
2. if `SIGNATURE_SPOOFING` is set to `no` - the default case - then **_only_** the LOS changes are applied, none of our patches
3. if `SIGNATURE_SPOOFING` is set to `restricted`, then a modified version of our `frameworks_base_patch` is applied, allowing only privileged apps to spoof signature
4. if `SIGNATURE_SPOOFING` is set to `yes`, then both our `frameworks_base_patch` and our `modules_permission_patch` are applied allowing any app to spoof signature

Our current build process for official (`userdebug`) builds now has `SIGNATURE_SPOOFING` is set to `no`, effectively allowing `restricted` spoofing using **_only_** the LOS changes and **_not using_** our own patches. I would like that logic to continue to work.

My preference would be to ***only*** apply the new patch if `BUILD_TYPE` ***is not*** `userdebug`: the patch is only required in this case

petefoth commented:
> My preference would be to _**only**_ apply the new patch if `BUILD_TYPE` _**is not**_ `userdebug`: the patch is only required in this case

A minimal level of testing would therefore include running a 21.0 build  with `SIGNATURE_SPOOFING` set to `no`, and checking the new patch ***is not*** applied. 

@nnsee As a matter of interest, have you made changes to the `lemonadep` sources allowing 21.0 builds to succeed? As far as I know all 21.0 builds (except for `nio`) are broken

nnsee commented:
That's more or less with what I'm currently writing, although instead of heuristically looking at whether the build variant is `userdebug` or not, I've simply introduced a new env var `USER_BUILD_SPOOFING` which, when set to "yes", applies my patch (by default, I've left it as "no"). Is this okay, or would you rather disregard the env var and that the patch is always applied if the build variant isn't `userdebug` or `eng`?

I'm guessing that they consider signature spoofing a "security issue" and this is why they only enable it in "debuggable" builds, and I'd like to leave the choice to entirely disable it in _user_ builds to the user if they wish to do so.

I literally just built 21 for `lemonadep` from the master branch and it worked fine with just the changes in this PR applied, no other changes.

petefoth commented:
> I've simply introduced a new env var `USER_BUILD_SPOOFING` which, when set to "yes", applies my patch (by default, I've left it as "no"). Is this okay,

That works (I think 😄 ). I guess the logic needs to be dependent on branch: your patch won't work for builds before 18.1 as the LOS code you're patching doesn't exist in earlier branches. So we probably need a check in [this switch statement](https://github.com/lineageos4microg/docker-lineage-cicd/blob/aed32ded7991a512ddf7eba8a5afa1caa4f79f5a/src/build.sh#L151C1-L173C1): if `USER_BUILD_SPOOFING` is set `true` for earlier builds then  write a suitable message to the log (`"Branch $branch does not support USER_BUILD_SPOOFING"`) and EITHER set `USER_BUILD_SPOOFING` to `false` and continue OR exit 

> I literally just built 21 for `lemonadep` from the master branch and it worked fine with just the changes in this PR applied, no other changes.

I would love to know which other devices will actually build for 21.0: according to [`hudson`](https://github.com/LineageOS/hudson/blob/main/lineage-build-targets) all devices are disabled

nnsee commented:
I've:

* introduced a new environment variable `USER_BUILD_SPOOFING` (defaults to `no`) - when set to `yes`, this applies the patch to enable the `user` build patch
* removed support for patching 21 with the default microG patches (`SIGNATURE_SPOOFING` set to `yes` or `restricted`) - it still builds, but prints a warning and doesn't attempt to apply any microG patches
* updated the README to reflect this

Anything I should change?

As an aside, I noticed that the `build.sh` script seems to be turning rather... spaghetty-ish. I'd probably introduce a future PR to try and organize it a bit and reduce the amount of nesting going on.

petefoth commented:
I'm not going to merge this directly to `master`. I've created a branch ([nns-v21-spoofing-patches](https://github.com/lineageos4microg/docker-lineage-cicd/tree/nns-v21-spoofing-patches)) that we can use for testing, and changed this PR. I'll create a new PR to merge to master when we're happy the changes are good 😄

petefoth commented:
OK a docker image based on this branch is now [available at DockerHub](https://hub.docker.com/layers/lineageos4microg/docker-lineage-cicd/nns-v21-spoofing-patches/images/sha256-1a51e7a8ba25f620ecd214158a0ba1d58aa2c87d59f3d81c97b45f95afa29238?context=explore) for testing
`docker pull lineageos4microg/docker-lineage-cicd:nns-v21-spoofing-patches`

petefoth commented:
> As an aside, I noticed that the `build.sh` script seems to be turning rather... spaghetty-ish. I'd probably introduce a future PR to try and organize it a bit and reduce the amount of nesting going on.

I take your point :) A lot of the nesting was introduced in implementing  "Make the operations performed by the docker image 'switchable'" #508 : almost by definition, optional operations need to be inside a conditional statement, and the script loops over every branch and every device specified in the `docker run` command, which makes two more levels of nesting unavoidable.

If we were going to simplify and restructure the code, then I would rewrite the whole script in python, rather than continuing with shell scripts. BUT ,as stated [in the README.md](https://github.com/lineageos4microg/docker-lineage-cicd#project-status)

> The project is currently in a fairly stable state:
> 
> -   we are (mostly) achieving our objective of delivering monthly builds
>     the only essential work that is ongoing is to
>     -    monitor the delivery process, to fix any problems that may occur, and to make any changes that are needed to ensure that the problems do not recur
>     -    to make any changes needed when upstreams make changes. In particular, when LineageOS introduces support for a new Android version and / or drops support for older Android versions
> 
> The project is therefore - in the opinion of the currently active maintainers - essentially 'feature complete' and in 'maintenance' mode. The only change that we believe might significantly improve the project is to support other classes of Android devices,

My opinion is that restructuring and improving the current code is _**not worth the effort**_ that would be involved in designing, implementing, testing, and documenting the necessary changes. There was [an attempt in 2022 to simplify and improve the functionality of `init.sh`](https://github.com/lineageos4microg/docker-lineage-cicd/pull/342). Work on the issue went o for over three months, and was eventually abandoned. 

Don't let that stop you from trying if it interests you, but there would be a lot of less interesting work to do (testing, documentation, reviewing etc.), some of which would have to be done by others (it's hard to review your own code 😄) before I would consider replacing the working code that we already have. I honestly think there are more productive things that you could do with your time and development skills.

petefoth commented:
> > OK a docker image based on this branch is now [available at DockerHub](https://hub.docker.com/layers/lineageos4microg/docker-lineage-cicd/nns-v21-spoofing-patches/images/sha256-1a51e7a8ba25f620ecd214158a0ba1d58aa2c87d59f3d81c97b45f95afa29238?context=explore) for testing `docker pull lineageos4microg/docker-lineage-cicd:nns-v21-spoofing-patches`
> 
> I decided to test this, but I'm getting this output unfortunately:
> 
> ```
> <snip>
> /root/build.sh: line 299: frameworks_base_patch: unbound variable
> ```
> 
> It seems that the warning I'm supposed to get here isn't working properly. It seems the check itself, which should check if the variable exists creates this crash itself. Only completely removing SIGNATURE_SPOOFING in my command fixes this (at least the build is running now, whether it fully passes will have to wait).

This has been addressed in https://github.com/lineageos4microg/docker-lineage-cicd/pull/612, but PR probably won;t get merged - see the following  comments
- https://github.com/lineageos4microg/docker-lineage-cicd/pull/612#discussion_r1575611381 
- https://github.com/lineageos4microg/docker-lineage-cicd/issues/607#issuecomment-2070874499
- https://github.com/lineageos4microg/docker-lineage-cicd/issues/607#issuecomment-2071356884

> 
> Personally, I think we should maintain the distinction between "yes" and "restricted". Some people want the flexibility of yes (and accept the risk, which is just that you could accidentally give it to an app you should not trust.). Others want the restriction of this permission for system apps. Perhaps we should look to keep this option possible. 

I agree, which is why I have separated the problem into two parts, which will (with luck) be fixed separately
- in #607 we'll try to get our patches working
- in #614 we'll try to make the upstream patches work for `user` builds. The fix for that will probably involve the changes made around the new `USER_BUILD_SPOOFING` flag

I hope that clarifies things a bit
